proxy/fdleak_test.go

```diff
@@ -0,0 +1,499 @@
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strconv"
+	"syscall"
+	"testing"
+)
+
+func fcntl(fd int, cmd int, arg int) (val int, err error) {
+	r0, _, e1 := syscall.Syscall(
+		syscall.SYS_FCNTL, uintptr(fd), uintptr(cmd), uintptr(arg))
+	val = int(r0)
+	if e1 != 0 {
+		err = e1
+	}
+	return
+}
+
+// FdLeakDetector detects fd leaks by letting its user take snapshots of the
+// list of opened fds at key points in the profiled program and compare two
+// snapshots.
+type FdLeakDetector struct {
+}
+
+type FdInfo struct {
+	Fd          int
+	Flags       int
+	CloseOnExec bool
+	Text        string
+}
+
+func (info *FdInfo) dump(w io.Writer) {
+	flags := ""
+	if info.CloseOnExec {
+		flags += "cloexec "
+	}
+
+	// file status
+	if info.Flags&syscall.O_APPEND != 0 {
+		flags += "append "
+	}
+	if info.Flags&syscall.O_NONBLOCK != 0 {
+		flags += "nonblock "
+	}
+
+	// acc mode
+	if info.Flags&syscall.O_RDONLY != 0 {
+		flags += "read-only "
+	}
+	if info.Flags&syscall.O_RDWR != 0 {
+		flags += "read-write "
+	}
+	if info.Flags&syscall.O_WRONLY != 0 {
+		flags += "write-only "
+	}
+	if info.Flags&syscall.O_DSYNC != 0 {
+		flags += "dsync "
+	}
+	if info.Flags&syscall.O_RSYNC != 0 {
+		flags += "rsync "
+	}
+	if info.Flags&syscall.O_SYNC != 0 {
+		flags += "sync "
+	}
+
+	fmt.Fprintf(w, "  %d: %s\n", info.Fd, info.Text)
+	fmt.Fprintf(w, "   : %s\n", flags)
```

Comment:
niggle - you'll end up with a trailing space here :)

It's not going to block this PR, but you could maybe make `flags` an array and then `Join` the values to avoid the easy-to-forget-to-add trailing spaces in the string values above).